pkg/mimo/actuator/service.go

```diff
@@ -0,0 +1,316 @@
+}
+
+type service struct {
+	dialer  proxy.Dialer
+	baseLog *logrus.Entry
+	env     env.Interface
+
+	dbGroup actuatorDBs
+
+	m        metrics.Emitter
+	mu       sync.RWMutex
+	cond     *sync.Cond
+	stopping *atomic.Bool
+	workers  *atomic.Int32
+
+	b buckets.BucketWorker
+
+	lastChangefeed atomic.Value //time.Time
+	startTime      time.Time
+
+	pollTime time.Duration
+	now      func() time.Time
+
+	tasks map[string]tasks.MaintenanceTask
+
+	serveHealthz bool
+}
+
+type actuatorDBs interface {
+	database.DatabaseGroupWithOpenShiftClusters
+	database.DatabaseGroupWithMaintenanceManifests
+}
+
+func NewService(env env.Interface, log *logrus.Entry, dialer proxy.Dialer, dbg actuatorDBs, m metrics.Emitter) *service {
+	s := &service{
+		env:     env,
+		baseLog: log,
+		dialer:  dialer,
+
+		dbGroup: dbg,
+
+		m:        m,
+		stopping: &atomic.Bool{},
+		workers:  &atomic.Int32{},
+
+		startTime: time.Now(),
+		now:       time.Now,
+		pollTime:  time.Minute,
+
+		serveHealthz: true,
+	}
+
+	s.b = buckets.NewBucketWorker(log, s.worker, &s.mu)
+
+	return s
+}
+
+func (s *service) SetMaintenanceTasks(tasks map[string]tasks.MaintenanceTask) {
+	s.tasks = tasks
+}
+
+func (s *service) Run(ctx context.Context, stop <-chan struct{}, done chan<- struct{}) error {
+	defer recover.Panic(s.baseLog)
+
+	// Only enable the healthz endpoint if configured (disabled in unit tests)
+	if s.serveHealthz {
+		c := &healthz.Handler{
+			Checks: map[string]healthz.Checker{
+				"ready": func(h *http.Request) error {
+					if !s.checkReady() {
+						return errors.New("not ready")
+					}
+					return nil
+				},
+			},
+		}
+
+		m := http.NewServeMux()
+		m.Handle("/healthz", http.StripPrefix("/healthz", c))
+		// Handle healthz subpaths
+		m.Handle("/healthz/", http.StripPrefix("/healthz", c))
+
+		h := &http.Server{
+			Handler:     m,
+			ErrorLog:    log.New(s.baseLog.Writer(), "", 0),
+			BaseContext: func(net.Listener) context.Context { return ctx },
+		}
+
+		listener, err := s.env.Listen()
+		if err != nil {
+			return err
+		}
+
+		go func() {
+			err := h.Serve(listener)
+			if err != http.ErrServerClosed {
+				s.baseLog.Error(err)
+			}
+		}()
+	}
+
+	t := time.NewTicker(10 * time.Second)
+	defer t.Stop()
+
+	if stop != nil {
+		go func() {
+			defer recover.Panic(s.baseLog)
+
+			<-stop
+			s.baseLog.Print("stopping")
+			s.stopping.Store(true)
+			s.cond.Signal()
+		}()
+	}
+	go heartbeat.EmitHeartbeat(s.baseLog, s.m, "actuator.heartbeat", nil, s.checkReady)
+
+	lastGotDocs := make(map[string]*api.OpenShiftClusterDocument)
+	for {
+		if s.stopping.Load() {
+			break
+		}
+
+		old, err := s.poll(ctx, lastGotDocs)
+		if err != nil {
+			s.baseLog.Error(err)
+		} else {
+			lastGotDocs = old
+		}
+
+		<-t.C
+	}
+
+	if !s.env.FeatureIsSet(env.FeatureDisableReadinessDelay) {
+		s.waitForWorkerCompletion()
+	}
+	s.baseLog.Print("exiting")
+	close(done)
+	return nil
+}
+
+// Temporary method of updating without the changefeed -- the reason why is
```

Comment:
"Temporary" as in "this code will eventually be replaced"? The "reason why is complicated" part makes me curious 😅

Reply:
Yes https://issues.redhat.com/browse/ARO-7974